### PR TITLE
Refactor mixins to inject {calibrator,selector,producer} info to tasks

### PR DESCRIPTION
**Pull request by riga**
This PR contains some structural refactoring. Through the `{Calibrator,Selector,Producer}Mixin` classes, tasks obtain the feature to call arbitrary array functions, configured via `--{calibrator(s),selector,producer(s)}` arguments. However, to properly integrate these externalized functions in a plug-and-play fashion, information such as required / produced systematic shifts need to be correctly propagated. This happens mostly in class methods of the base tasks, since these decisions need to be taken _before_ a task class is instantiated.

The PR refactors

- how task classes can define their shifts (they can now depend on the above mentioned array functions, _plus_ they allow resolving this same information for dependent tasks), and
- how array functions can customize used / produced columns, as well as systematic shifts (they can wrap a function accepting analysis, config, dataset inst's, and performing updates to (e.g.) their shifts).

### CODE REVIEW

File: ap/columnar_util.py
Comment (dsavoiu) [inline ap/columnar_util.py:754]: AFAIU the `shifts` keyword currently only declares dependent shifts (as in the shifts that affect the result of the array function and for which the column aliases should be applied before the array function is executed). Wouldn't it make sense to also declare produced shifts via a separate keyword?

File: ap/columnar_util.py
Comment (dsavoiu) [inline ap/columnar_util.py:895]: I think one should also run over `self.shifts` here. Right now if other array functions are specified via `shifts` the `update` function is not run on them, which may lead to wrong `shift` resolution. One workaround would be to add these to `uses`, but if one is just interested in the shifts and not the results, passing them via `shifts` would be semantically cleaner. 
See also PR #35 and in particular this commit https://github.com/uhh-cms/analysis_playground/pull/35/commits/607e534fec9c6bcc407afcafc984f25560f349c7 for an example where this becomes relevant.

File: ap/columnar_util.py
Comment (dsavoiu) [inline ap/columnar_util.py:974]: Also run over `self.shifts` (see comment for `run_update`)

File: ap/columnar_util.py
Comment (dsavoiu) [inline ap/columnar_util.py:925]: Also propagate to `self.shifts` (see comment for `run_update`)

File: ap/columnar_util.py
Comment (dsavoiu) [inline ap/columnar_util.py:862]: Shoudn't the line above be `shifts |= obj.all_shifts` to properly resolve all dependent shifts?

File: ap/columnar_util.py
Comment (dsavoiu) [inline ap/columnar_util.py:855]: Also propagate to `self.shifts` (see comment for `run_update`)

File: ap/tasks/selection.py
Comment (dsavoiu) [inline ap/tasks/selection.py:23]: This should probably be removed and injected dynamically by the JEC calibrator (already done in PR #35)

**Review by dsavoiu — APPROVED** [review]
This is very nice! I played around a lot with this while implementing JEC/JER and it looks like everything is mostly working as expected, with the exception of some use cases where shifts don't seem to propagate properly to dependent array functions, or at least not as I would expect them to :) I've added some comments with more details at the relevant lines in the code.

I'm marking this as approved, but maybe you could also have a look at PR #35 before merging, since I have included these changes there as well and also added some commits that I believe solve the shift propagation issues I was seeing. If you agree with them there should be no need to further change this PR and it can be merged together with PR #35. In case there are any issues feel free to follow up here or there, otherwise everything looks good on my end.

File: ap/columnar_util.py
Comment (riga) [reply ap/columnar_util.py:754]: I also thought about that :)
But as far as the ShiftTask mechanism is concerned, dependent and produced shifts don't have a special distinction. For this matter, running with a certain shifts only means that "the output will be different than for 'nominal'".

File: ap/columnar_util.py
Comment (riga) [reply ap/columnar_util.py:895]: True! I'll address this in the next commit.

File: ap/columnar_util.py
Comment (riga) [reply ap/columnar_util.py:974]: True! I'll address this in the next commit.

File: ap/columnar_util.py
Comment (riga) [reply ap/columnar_util.py:925]: True! I'll address this in the next commit.

File: ap/columnar_util.py
Comment (riga) [reply ap/columnar_util.py:862]: Yes it should. And thinking about it, one shouldn't loop here over the `uses` and `produces` sets, but just look for other `ArrayFunction`s in `self.shifts`. This requires adding a dependent function in (e.g.) both `produces` _and_ `shifts`, but at the same time provides a better level of configurability.

File: ap/columnar_util.py
Comment (riga) [reply ap/columnar_util.py:855]: ✔︎

File: ap/tasks/selection.py
Comment (riga) [reply ap/tasks/selection.py:23]: Yep. I would leave it to #35 to remove that :)

File: ap/columnar_util.py
Comment (riga) [reply ap/columnar_util.py:869]: Thinking about this further, we should leave these special cases for the future and only take action if required.

File: ap/columnar_util.py
Comment (dsavoiu) [reply ap/columnar_util.py:760]: I see, thanks for clarifying. Guess I was just a bit confused by the formulation in the docstring, but when you put it that way it makes sense.

File: ap/columnar_util.py
Comment (dsavoiu) [reply ap/columnar_util.py:860]: 👍

File: ap/columnar_util.py
Comment (dsavoiu) [reply ap/columnar_util.py]: Thanks! Looks good now.

File: ap/columnar_util.py
Comment (dsavoiu) [reply ap/columnar_util.py:869]: Hm, I think it actually makes sense to also loop over `produces` and `uses` here. To my understanding, one would always want the `shifts` from these dependencies to be propagated to the current function's `all_shifts`, since they indirectly affect the result of the current function as well.

I would add them for now, but we should keep this in mind in case we need to make this more flexible for some future use case.

**Comment by dsavoiu** [comment]
Looks good now, going ahead with the merge.

**Pull request merged**
